### PR TITLE
Add support for logging custom events

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Adds support for logging Application Insights [CustomEvents](https://learn.microsoft.com/en-us/azure/azure-monitor/app/data-model-complete#event)

New API

`public void TrackEvent(string name, IReadOnlyList<KeyValuePair<string, object?>>? attributes = null)`

### Exporter users:

```csharp
this.loggerFactory = LoggerFactory.Create(builder =>
{
    builder.AddOpenTelemetry(options =>
    {
        options.SetResourceBuilder(resourceBuilder);
        options.AddAzureMonitorLogExporter(o => o.ConnectionString = connectionString, credential);
    });
});

var eventLogger = new ApplicationInsightsEventLogger(this.loggerFactory);

var attributesList = new List<KeyValuePair<string, object?>>
    {
        new KeyValuePair<string, object?>("Attribute1", "value1"),
        new KeyValuePair<string, object?>("Attribute2", "value2")
    };

eventLogger.TrackEvent("MyCustomEvent", attributesList);
```

### Distro users: 
CustomEventLogger will be auto-injected during the set up and will be available via DI

```csharp
var app = builder.Build();
app.MapGet("/", (IApplicationInsightsEventLogger ApplicationInsightsEventLogger) =>
{
    ApplicationInsightsEventLogger.TrackEvent("MyCustomEvent");
    .
    .
});
```

TODO: Add changelog entry and update Readme details.

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Azure.Monitor.OpenTelemetry.AspNetCore.csproj [hunk 1]
```diff
@@ -23,10 +23,10 @@
     <!-- Depending on monthly deliverables, we may switch between PackageReference or ProjectReference. Keeping both here to make the switch easier. -->
 
     <!-- FOR PUBLIC RELEASES, MUST USE PackageReference. THIS REQUIRES A STAGGERED RELEASE IF SHIPPING A NEW EXPORTER. -->
-    <PackageReference Include="Azure.Monitor.OpenTelemetry.Exporter" />
+    <!--<PackageReference Include="Azure.Monitor.OpenTelemetry.Exporter" />-->
 
     <!-- FOR LOCAL DEV, ProjectReference IS PREFERRED. -->
-    <!--<ProjectReference Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Azure.Monitor.OpenTelemetry.Exporter.csproj" />-->
+    <ProjectReference Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Azure.Monitor.OpenTelemetry.Exporter.csproj" />
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Azure.Monitor.OpenTelemetry.AspNetCore.csproj:29]: Please undo the changes for this project.

Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Azure.Monitor.OpenTelemetry.AspNetCore.csproj:29]: This is needed until we release the exporter package.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/api/Azure.Monitor.OpenTelemetry.Exporter.net6.0.cs [hunk 2]
```diff
@@ -1,3 +1,15 @@
+namespace Azure.Monitor.OpenTelemetry.CustomEvents
+{
+    public sealed partial class CustomEventLogger : Azure.Monitor.OpenTelemetry.CustomEvents.ICustomEventLogger
+    {
+        public CustomEventLogger(Microsoft.Extensions.Logging.ILoggerFactory loggerFactory) { }
+        public void TrackEvent(string name, System.Collections.Generic.IReadOnlyList<System.Collections.Generic.KeyValuePair<string, object?>>? attributes = null) { }
+    }
+    public partial interface ICustomEventLogger
+    {
+        void TrackEvent(string name, System.Collections.Generic.IReadOnlyList<System.Collections.Generic.KeyValuePair<string, object?>>? attributes = null);
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/api/Azure.Monitor.OpenTelemetry.Exporter.net6.0.cs:10]: Earlier API's did not support nullable, not sure it has changed. Please verify.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/CustomEventLogger.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.Logging;
+
+namespace Azure.Monitor.OpenTelemetry.CustomEvents
+{
+    /// <summary>
+    /// Default logger for logging custom events telemetry.
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/CustomEventLogger.cs:11]: nit: Default logger for logging application insights custom events telemetry.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/CustomEventLogger.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            if (loggerFactory == null)
+                throw new ArgumentNullException(nameof(loggerFactory));
+
+            _logger = loggerFactory.CreateLogger("Azure.Monitor.OpenTelemetry.CustomEvents");
+        }
+
+        /// <inheritdoc/>
+        public void TrackEvent(string name, IReadOnlyList<KeyValuePair<string, object?>>? attributes = null)
+        {
+            // Setting the loglevel to critical to ensure events are not filtered.
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/CustomEventLogger.cs:38]: Additionally specifying the loglevel is not honored for custom events in the comments would help.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/CustomEventLogger.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        private static readonly Func<IReadOnlyList<KeyValuePair<string, object?>>?, Exception?, string> s_formatter = (state, ex) =>
+        {
+            return "CustomEvent";
+        };
+
+        private readonly ILogger _logger;
+
+        /// <summary>
+        /// Custom events.
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/CustomEventLogger.cs:23]:
```suggestion
        /// creates an instance of CustomEventLogger to log custom events.
```

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ICustomEventLogger.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Azure.Monitor.OpenTelemetry.CustomEvents
+{
+    /// <summary>
+    /// An interface for logging custom events telemetry.
+    /// </summary>
+    public interface ICustomEventLogger
+    {
+        /// <summary>
+        /// Tracks custom events.
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ICustomEventLogger.cs:14]:
```suggestion
        /// Tracks Application Insights custom events.
```

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs [hunk 7]
```diff
@@ -56,17 +56,27 @@ internal static List<TelemetryItem> OtelToAzureMonitorLogs(Batch<LogRecord> batc
...
                     if (logRecord.Exception != null)
                     {
+                        telemetryItem = new TelemetryItem("Exception", logRecord, resource, instrumentationKey);
                         telemetryItem.Data = new MonitorBase
                         {
                             BaseType = "ExceptionData",
                             BaseData = new TelemetryExceptionData(Version, logRecord),
                         };
                     }
+                    else if (logRecord.CategoryName == "Azure.Monitor.OpenTelemetry.CustomEvents")
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs:68]: can we use referenceequals as we use this string within the same library?

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs [hunk 8]
```diff
@@ -56,17 +56,27 @@ internal static List<TelemetryItem> OtelToAzureMonitorLogs(Batch<LogRecord> batc
...
+                        telemetryItem = new TelemetryItem("Event", logRecord, resource, instrumentationKey);
+                        telemetryItem.Data = new MonitorBase
+                        {
+                            BaseType = "EventData",
+                            BaseData = new TelemetryEventData(Version, logRecord),
+                        };
+                    }
                     else
                     {
+                        telemetryItem = new TelemetryItem("message", logRecord, resource, instrumentationKey);
```
[the remark this answers is not in the document]
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs]: Yeah, that was a typo. fixed.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/api/Azure.Monitor.OpenTelemetry.Exporter.net6.0.cs (hunk 2 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/api/Azure.Monitor.OpenTelemetry.Exporter.net6.0.cs:10]: We had defaults: https://learn.microsoft.com/en-us/dotnet/api/microsoft.applicationinsights.telemetryclient.trackevent?view=azure-dotnet#microsoft-applicationinsights-telemetryclient-trackevent(system-string-system-collections-generic-idictionary((system-string-system-string))-system-collections-generic-idictionary((system-string-system-double)))

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/CustomEventLogger.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.Logging;
+
+namespace Azure.Monitor.OpenTelemetry.Events
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/CustomEventLogger.cs:8]: Just curious, how did we land on this namespace?

It's unusual because it doesn't match the project & directory (Azure.Monitor.OpenTelemetry.Exporter.CustomEvents).
Second, all our other public classes are in the root directory (Azure.Monitor.OpenTelemetry.Exporter).

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs (hunk 7 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/LogsHelper.cs:68]: My understanding is it uses that by default.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/api/Azure.Monitor.OpenTelemetry.Exporter.net6.0.cs [hunk 10]
```diff
@@ -1,3 +1,15 @@
+namespace Azure.Monitor.OpenTelemetry.CustomEvents
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/api/Azure.Monitor.OpenTelemetry.Exporter.net6.0.cs:1]: This namespace doesn't match the one used in the classes.
Is this an incomplete refactor?

Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/api/Azure.Monitor.OpenTelemetry.Exporter.net6.0.cs]: yes, fixed.

**Comment by TimothyMothra** [comment]
TODO: Need to update Integration Tests. Can be done as a separate PR. Likely, this will be my work item :)

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs [hunk 11]
```diff
@@ -544,6 +545,45 @@ public void DuplicateKeysInLogRecordAttributesAndLogScope2()
...
+            eventLogger.TrackEvent("MyCustomEvent", attributesList);
+
+            // Assert.
+            var logRecord = logRecords.Single();
+            var logRecodBatch = new Batch<LogRecord>(new[] { logRecord } , 1);
+            var properties = new ChangeTrackingDictionary<string, string>();
+            var telemetryItems = LogsHelper.OtelToAzureMonitorLogs(logRecodBatch, null, "Ikey");
+
+            var telemetryItem = telemetryItems.FirstOrDefault();
+            var telemetryEventData = telemetryItem?.Data.BaseData as TelemetryEventData;
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs:577]: Using `FirstOrDefault` doesn't help here and may make it harder to troubleshoot test failures.

```suggestion
            var telemetryItem = telemetryItems.Single();
            var telemetryEventData = telemetryItem.Data.BaseData as TelemetryEventData;
```

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/api/Azure.Monitor.OpenTelemetry.Exporter.net6.0.cs (hunk 2 above)
Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/api/Azure.Monitor.OpenTelemetry.Exporter.net6.0.cs:10]: I meant the Azure SDK API's.

**Comment by TimothyMothra** [comment]
TODO: please update changelogs for both Exporter and Distro

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/CustomEventLogger.cs (hunk 9 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/CustomEventLogger.cs:8]: It is going to be used in both distro and exporter so wanted something common. I am not sure if namespace matching the directory/project is a requirement.  Found at least one similar example: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/storage/Azure.Storage.Blobs.Batch/api/Azure.Storage.Blobs.Batch.netstandard2.0.cs

**Review by rajkumar-rangaraj — APPROVED** [review]

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        /// Creates an instance of CustomEventLogger to log custom events.
+        /// </summary>
+        /// <param name="loggerFactory">LoggerFactory instance for logging events.</param>
+        /// <exception cref="ArgumentNullException"></exception>
+        public ApplicationInsightsEventLogger(ILoggerFactory loggerFactory)
+        {
+            if (loggerFactory == null)
+                throw new ArgumentNullException(nameof(loggerFactory));
+
+            _logger = loggerFactory.CreateLogger("Azure.Monitor.OpenTelemetry.CustomEvents");
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:32]: Move to a constant somewhere.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.Logging;
+
+namespace Azure.Monitor.OpenTelemetry.Events
+{
+    /// <summary>
+    /// Default logger for logging Application Insights custom events.
+    /// </summary>
+    public sealed class ApplicationInsightsEventLogger : IApplicationInsightsEventLogger
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: Users typically have other logger providers registered as well. Can you comment how this events will look like in the other providers? Since `Critical` severity is used, it'd start showing up by default in other places.

Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: > Can you comment how this events will look like in the other providers

Are you asking to clarify that in summary?. They will just look like any other log with `critical` severity.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/api/Azure.Monitor.OpenTelemetry.Exporter.net6.0.cs (hunk 2 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/api/Azure.Monitor.OpenTelemetry.Exporter.net6.0.cs:10]: I added a todo for following up on this.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs (hunk 13 above)
Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: I am not sure yet... People may get shocked seeing Critical logs all of a sudden showing up in their other LogProviders (like ConsoleProvider which is enabled by default).

Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: Agree, that is an issue. An alternate option is to use informational and document that users need to allow informational logs from `Azure.Monitor.OpenTelemetry.CustomEvents` category.

Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: @vishweshbankwar what if the Exporter creates an extra LoggerFactory that only sends CustomEvents to Azure Monitor?

Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: ^ That won't meet our needs, as it won't play with LogRecordProcessor and other things in the pipeline, so existing pipeline (LoggerFactory) should be used

Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: _An alternate option is to use informational and document that users need to allow informational logs from Azure.Monitor.OpenTelemetry.CustomEvents category (that is not perfect either)._

Can we inject such a filter automatically?

Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: No, we need access to IloggingBuilder for that.

Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: don't we get access to it from the service collection ?

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: If we rely on service collection, classic .NET framework may not be able to take advantage of this feature. Based on our understanding customEvents are used mostly in classic applications.

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: > An alternate option is to use informational and document that users need to allow informational logs from Azure.Monitor.OpenTelemetry.CustomEvents category (that is not perfect either).

We could achieve this with the distro and the anticipated new UseAzureMonitorExporter API.

Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:13]: I changed it to information. Applied the filters for distro users. Will add documentation for users creating the logger instance manually.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        private const string EventLoggerName = "Azure.Monitor.OpenTelemetry.CustomEvents";
+        private static readonly Func<IReadOnlyList<KeyValuePair<string, object?>>?, Exception?, string> s_formatter = (state, ex) =>
+        {
+            return "CustomEvent";
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/ApplicationInsightsEventLogger.cs:18]: use more detailed name, so folks using other log providers will not get confused seeing this..

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs [hunk 15]
```diff
@@ -544,6 +545,45 @@ public void DuplicateKeysInLogRecordAttributesAndLogScope2()
...
+        [Fact]
+        public void ValidateCustomEvent()
+        {
+            // Arrange.
+            var logRecords = new List<LogRecord>(1);
+            using var loggerFactory = LoggerFactory.Create(builder =>
+            {
+                builder.AddOpenTelemetry(options =>
+                {
+                    options.AddInMemoryExporter(logRecords);
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs:557]: not really necessary, but could you add a test with a `LogRecordProcessor` and see that it can inspect this, just like any other normal ILogger logs.. (it should be, but i am just making sure everything is covered!)

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs [hunk 16]
```diff
@@ -211,6 +213,10 @@ public static OpenTelemetryBuilder UseAzureMonitor(this OpenTelemetryBuilder bui
                     }
                 });
 
+            // Enable custom events and set filter to enable collection.
+            builder.Services.AddSingleton<IApplicationInsightsEventLogger, ApplicationInsightsEventLogger>();
+            builder.Services.AddLogging(o => o.AddFilter(EventLoggerName, LogLevel.Information));
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs:218]: Could you validate with unittest that user can override this filter, to achieve the effect of disabling EventTelemetry completely, if they chose to?

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs [hunk 17]
```diff
@@ -211,6 +213,34 @@ public static OpenTelemetryBuilder UseAzureMonitor(this OpenTelemetryBuilder bui
                     }
                 });
 
+            // Enable custom events and set filter to enable collection.
+            builder.Services.AddSingleton<IApplicationInsightsEventLogger, ApplicationInsightsEventLogger>();
```
Comment (CodeBlanch) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs:217]: nit: Call `TryAddSingleton`

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/IApplicationInsightsEventLogger.cs [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+    /// </summary>
+    public interface IApplicationInsightsEventLogger
+    {
+        /// <summary>
+        /// Tracks Application Insights custom events.
+        /// </summary>
+        /// <param name="name">Name of the CustomEvent.</param>
+        /// <param name="attributes">Custom dimensions of the event.</param>
+        public void TrackEvent(string name, IReadOnlyList<KeyValuePair<string, object?>>? attributes = null);
+    }
```
Comment (CodeBlanch) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/IApplicationInsightsEventLogger.cs:19]: Some other possible shapes to consider:

```csharp
public void TrackEvent(string name, in TagList attributes = default);
public void TrackEvent(string name, params ReadOnlySpan<KeyValuePair<string, object?>> attributes); // May be hard/impossible to implement with ILogger or require newer runtime features but would be awesome to support :)
```

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs [hunk 19]
```diff
@@ -211,6 +213,34 @@ public static OpenTelemetryBuilder UseAzureMonitor(this OpenTelemetryBuilder bui
...
+            //   ...
+            // }
+            // The reason is as both rules will match the filter, the last one added wins.
+            // To ensure that the default filter is in the beginning of filter rules, so that user override from Configuration will always win,
+            // we add code filter rule to the 0th position as below.
+            builder.Services.Configure<LoggerFilterOptions>(
+                options => options.Rules.Insert(
+                    0,
+                    new LoggerFilterRule(
+                        "OpenTelemetry.Logs.OpenTelemetryLoggerProvider",
```
Comment (CodeBlanch) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs:239]: Just peaked at the docs for `LoggerFilterRule` it seems you can specify [provider type or alias](https://learn.microsoft.com/dotnet/api/microsoft.extensions.logging.loggerfilterrule.providername). The alias we use for OTel is "OpenTelemetry" (defined [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/1b3f1894d95fd01fe475245ec60ef273d5f77bb6/src/OpenTelemetry/Logs/ILogger/OpenTelemetryLoggerProvider.cs#L16)) should we do that here instead? Just thinking the alias would be nice to not tie this code to the implementation SDK happens to be using.

**Review by CodeBlanch — APPROVED** [review]
Sent some comments but LGTM

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs (hunk 19 above)
Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs:239]: +1

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryEventData.cs [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+using OpenTelemetry.Logs;
+
+namespace Azure.Monitor.OpenTelemetry.Exporter.Models
+{
+    internal partial class TelemetryEventData
+    {
+        public TelemetryEventData(int version, LogRecord logRecord) : base(version)
+        {
+            Properties = new ChangeTrackingDictionary<string, string>();
+            Measurements = new ChangeTrackingDictionary<string, double>();
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryEventData.cs:13]: a separate issue: Given there is no way user can populate Measurements, any way we can avoid newing up this Dictionary? I am assuming we are doing it for each and every telemetry only to just throw it away.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryEventData.cs [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+
+            Name = logRecord.EventId.Name;
+
+            if (logRecord.Attributes != null)
+            {
+                foreach (var kv in logRecord.Attributes)
+                {
+                    Properties.Add(kv.Key, kv.Value?.ToString());
```
Comment (cijothomas) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryEventData.cs:21]: if we are stringifying here anyway, could the API just accept kvp of string,string, instead of string,object?

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryEventData.cs (hunk 20 above)
Comment (TimothyMothra) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryEventData.cs:13]: +1 to this, but would need to confirm that change doesn't break the generated serialization code. ie: if generated code expects a not-null value.

Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryEventData.cs:13]: IIRC, null value causes an issue with serialization. That is why we have this in all telemetry types.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs (hunk 19 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs]: updated

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryEventData.cs (hunk 21 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryEventData.cs:21]: Sdk wouldn't be able to detect the state in that case: https://github.com/open-telemetry/opentelemetry-dotnet/blob/1b3f1894d95fd01fe475245ec60ef273d5f77bb6/src/OpenTelemetry/Logs/ILogger/OpenTelemetryLogger.cs#L166-L175

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs (hunk 15 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/tests/Azure.Monitor.OpenTelemetry.Exporter.Tests/LogsHelperTests.cs:557]: its for creating the batch later.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/IApplicationInsightsEventLogger.cs (hunk 18 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/CustomEvents/IApplicationInsightsEventLogger.cs:19]: Like the Taglist option. @rajkumar-rangaraj - let me know what you think, can change it as a follow up.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/AzureSdkLoggingTests.cs [hunk 22]
```diff
@@ -22,11 +22,7 @@ namespace Azure.Monitor.OpenTelemetry.AspNetCore.Tests
 {
     public class AzureSdkLoggingTests
     {
-#if NET6_0
-        [ConditionallySkipOSTheory(platformToSkip: "macos", reason: "This test consistently exceeds 1 hour runtime limit when running on MacOS & Net60")]
-#else
-        [Theory]
-#endif
+        [Theory(Skip = "Test is unstable and need to be re-written without the need to create a web server.")]
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/AzureSdkLoggingTests.cs:25]: these tests are failing because of name - Azure.Monitor.OpenTelemetry.CustomEvents. This needed to be excluded in the  LogForwarder

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/AzureSdkLoggingTests.cs:25]: It could be a follow up.

**Review by rajkumar-rangaraj — APPROVED** [review]

**Comment by TimothyMothra** [comment]
> TODO: please update changelogs for both Exporter and Distro

@vishweshbankwar can you please take care of this?

**Comment by vishweshbankwar** [comment]
> > TODO: please update changelogs for both Exporter and Distro
> 
> @vishweshbankwar can you please take care of this?

Yes, I will be following up with this and Readme update

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryEventData.cs (hunk 20 above)
Comment (cijothomas) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Customizations/Models/TelemetryEventData.cs:13]: maybe create a dictionary once, and re-use throughout to avoid null issues. and save on the large number of unwanted heap allocations.
